### ISSUE TITLE
Atom consistently failing first install attempt

### ISSUE THREAD

**Issue by wbt** [atom/atom#20299]
### Description
Atom consistently fails on the first installation attempt; rerunning the same installer again succeeds.

### Steps to Reproduce

1. On Win10 Pro, the only installed version of Atom (1.42.0) was uninstalled using the Programs and Features Uninstall option.  
2. Wait a few minutes to allow completion. 
3. The latest AtomSetup-x64.exe installer, previously downloaded from the release page for v1.42.0 (which is the latest available at time of writing), is triggered, just once.
4. The following installation failure message is shown:
![Install failed](https://user-images.githubusercontent.com/563406/71834772-fce2f900-307d-11ea-9907-10303f1d4c60.PNG)
5. Clicking Open Setup Log provides [no helpful information](https://github.com/atom/atom/issues/20297) and that file is closed.
6. Wait a few minutes to make sure everything is calmed down.
7. As the definition of strategic software debugging is apparently to try the same thing again expecting different results, repeat step 3, retriggering the same installer.
8. This time, the splash screen shows for a bit:
![Install splash screen](https://user-images.githubusercontent.com/563406/71834869-3d427700-307e-11ea-830c-a451da8b2412.PNG)
and then Atom launches. 
9. Wait a few minutes, then close Atom.
10. This entire cycle can be repeated several times.  However, restarting the computer after step 2 seems to cause steps 4-7 to be skipped. 

The real setup log, Squirrel-Install.log (discussion [here](https://github.com/atom/atom/issues/20297)) contains the following example record from the failed attempt:

[06/01/20 12:02:42] error: IEnableLogger: Failed to copy Update.exe to C:\Users\username\AppData\Local\atom\Update.exe: System.UnauthorizedAccessException: Access to the path 'C:\Users\username\AppData\Local\atom\Update.exe' is denied.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Squirrel.Update.Program.<>c__DisplayClass4_1.<Install>b__4()
   at Squirrel.Utility.LogIfThrows(IFullLogger This, LogLevel level, String message, Action block)
[06/01/20 12:02:42] fatal: Finished with unhandled exception: System.AggregateException: One or more errors occurred. ---> System.UnauthorizedAccessException: Access to the path 'C:\Users\username\AppData\Local\atom\Update.exe' is denied.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Squirrel.Update.Program.<>c__DisplayClass4_1.<Install>b__4()
   at Squirrel.Utility.LogIfThrows(IFullLogger This, LogLevel level, String message, Action block)
   at Squirrel.Utility.ErrorIfThrows(IEnableLogger This, Action block, String message)
   at Squirrel.Update.Program.<Install>d__4.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Squirrel.Update.Program.executeCommandLine(String[] args)
   at Squirrel.Update.Program.main(String[] args)
---> (Inner Exception #0) System.UnauthorizedAccessException: Access to the path 'C:\Users\username\AppData\Local\atom\Update.exe' is denied.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Squirrel.Update.Program.<>c__DisplayClass4_1.<Install>b__4()
   at Squirrel.Utility.LogIfThrows(IFullLogger This, LogLevel level, String message, Action block)
   at Squirrel.Utility.ErrorIfThrows(IEnableLogger This, Action block, String message)
   at Squirrel.Update.Program.<Install>d__4.MoveNext()<---

**Expected behavior:**
Atom installs successfully the first time, absent any observed or triggered changes in the environment between the first and second installation attempts.

**Reproduces how often:**
This reproduced very reliably through several loops, prior to adding the computer restart step to the cycle.  After that it appears hard to reproduce, which seems likely to be a big hindrance to debugging the issue.  However, I am still reporting it in case others encounter a similar problem.

### Versions
1.42.0

### Additional Information
This does **not** appear to be a duplicate of [20275 ("updating to v1.42.0 faild on windows")](https://github.com/atom/atom/issues/20275) where a different root cause is reported in logs.

**Comment by WilSnoop** [comment]
> Access to the path 'C:\Users\username\AppData\Local\atom\Update.exe' is denied

At some point my Win10 was marking random folders as privileged. My problem was resolved with a Windows update. Somehow I doubt that this is your problem, however see if you can access the folder from File Explorer.

Access may be denied to a file if the system 'thinks' you are accessing a file already. I have seen this happen before. The fastest remedy is to log out and log back in. Try doing so if the installation was aborted or before uninstalling.

Is your user name in actual fact "username"? If it is something else: does your username contain non-ascii characters or spaces?

The fastest workaround for installation issues is to make a 'portable' installation:
 https://flight-manual.atom.io/getting-started/sections/installing-atom/#portable-mode

Any additional informations that you could provide would still be welcome.

**Comment by wbt** [comment]
>Is your user name in actual fact "username"? If it is something else: does your username contain non-ascii characters or spaces?  

Good question. The actual username for this install not in fact 'username' but you can treat it that way for the purpose of this issue.  What that is substituting for also contains only simple lowercase ASCII characters in the [a-z] range with no spaces and does not match any known or imaginably likely system keyword.  You can also treat the substitute value "username" as an upper bound on length. 

>Somehow I doubt that this is your problem, however see if you can access the folder from File Explorer.  
 
I can access the directory 'C:\Users\username\AppData\Local\atom\' and see a directory listing just fine.   I can also go into subdirectories with that same File Explorer window.

>My problem was resolved with a Windows update. ... The fastest remedy is to log out and log back in. Try doing so if the installation was aborted or before uninstalling....The fastest workaround for installation issues is to make a 'portable' installation

As reported above, simply re-running the same installer immediately after dismissing the failure message was an even faster workaround for this particular issue.  That also seems to rule out at least a fair chunk of other hypothesized explanations, possibly including the folder having been marked privileged by something other than Atom at random.  

>The fastest workaround for installation issues is to make a 'portable' installation:
https://flight-manual.atom.io/getting-started/sections/installing-atom/#portable-mode

While that may be true in general, a "portable" installation seems to foreclose system integration options such as the "Open with Atom" context menu item and Pin to Start.   I commented on that [here](https://github.com/atom/atom/issues/20296#issuecomment-571301003), disappointed in the design decision to never allow a user to specify the install path in the installer.

**Comment by WilSnoop** [comment]
I was hoping some more information came forward that would help the maintainers to diagnose this issue. In my view, there is not much to work with.

I think the key to this issue is with 'Squirrel' that handles the installation. Through the years of use of Atom, I have seen other (Windows) users have problems with 'Squirrel'.

Please confirm these:
+ Installation done with admin account or not.
+ Third party (internet) security software or anti-virus active.
+ Running Windows natively or in a virtual machine

---

> While that may be true in general, a "portable" installation seems to foreclose system integration options such as the "Open with Atom" context menu item and Pin to Start.

In the settings there is functionality provided to register the Atom application to use `open with` functionality. See "Settings -> System -> Register as file handler". There is even a "Show in file context menus". _My portable install for an Atom nightly version uses the last mentioned._

`Pin to Start` is available on the context menu. It can be triggered when right-clicking on the EXE file in File Explorer. _My portable install for an Atom nightly version is pinned to START and to the taskbar._

The portable installation does not update itself to a new version. Personally I am okay with this as this gives me control over when a version is updated.

**Comment by wbt** [comment]
- Admin account, though the installer does not request escalated/admin privileges    
- ESET Endpoint Antivirus running, and not going to be disabled for testing on this issue
- Windows running natively    

Thanks for the pointer to Settings->System.  I am disappointed to see that these settings do not carry over from one install to another, even following the [instructions](https://discuss.atom.io/t/error-while-installing-atom/69666/2) to preserve configuration that [you pointed me to](https://github.com/atom/atom/issues/20202#issuecomment-564200963).  This tells me there's no way to know just what settings will be preserved when trying to upgrade, which strongly disincentivizes upgrading for the purpose of having a slightly higher version number in bug reports on code that hasn't actually changed. 

`Pin to Start` is available on the context menu in File Explorer.  However, when I select this option, nothing happens.  Even if I drag the portable-install executable or a shortcut to it onto the start menu, it is not pinned to or even available in the Windows 10 start menu, except via search.  Some other "portable" apps have the same issue, and I think this is a bug in Windows (which is famously not open-source and not responsive to bug reports like that).  The only way I have found to get Pin to Start working in those cases is to use an installer / non-"portable" version of apps.  

Therefore, while suggesting an installer workaround of using the portable version may help some people by allowing most of the functionality to be gained in a crunch, it doesn't address this Issue or the other installer issues that I've recently raised, and is not even the best workaround for this specific issue (simply re-running the same installer is).  

If we're going to offer the installer option, it should work; if we can't make one that works with Squirrel we should drop it and either use a different installer-builder or provide clear instructions about how to use the portable version for everybody to achieve the same level of functionality.

**Comment by WilSnoop** [comment]
Carrying over the configuration can be done.
Pinning to START is no issue for me.

It is clear that you have your mind set on one way of working.
Efforts from a user is not what you want.

Good luck.
